Added support for build key in v3

surajnarwade opened the pull request:
Resolves #636 

This PR will add support for `build` in docker compose v3, as docker/cli#481 got merged now.
also this will update `sirupsen`, `docker/cli`, `docker/libcompose` in `glide`,
Also changed `Sirupsen` with `sirupsen` in all kompose packages as well as in
`docker/distribution` packages

cdrage commented:
@surajnarwade 
Please update commit description message (more detail, not just resolves X).

Tests would be good too!

cdrage commented:
Hey @surajnarwade I don't see why there is an issue with this. You are only updating Version 3 of Docker Compose which uses `docker/cli`, you shouldn't be touching / using `docker/libcompose`.

I also see that `docker/cli` is already using the most up-to-date logrus here: https://github.com/docker/cli/commit/e3b7700fb9ac7df392e4915978bfc714d9dacad8

What issues are you running into? Please investigate and let me know!

cdrage commented:
Also the error is just a simple renaming of `Sirupsen` in completion.go:

```
cmd/completion.go:9:2: cannot find package "github.com/Sirupsen/logrus" in any of:
	/home/travis/gopath/src/github.com/kubernetes/kompose/vendor/github.com/Sirupsen/logrus (vendor tree)
	/home/travis/.gimme/versions/go1.6.linux.amd64/src/github.com/Sirupsen/logrus (from $GOROOT)
	/home/travis/gopath/src/github.com/Sirupsen/logrus (from $GOPATH)
make: *** [test-unit-cover] Error 1
```

cdrage commented:
After the rename it seems to work:

```
github.com/kubernetes/kompose  master ✔                                                                                                                                                                                                                                  22h19m  
▶ git-pull-pr 846
Upstream:  kubernetes/kompose
Already on 'master'
error: branch 'pr_846' not found.
remote: Counting objects: 102, done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 102 (delta 55), reused 83 (delta 55), pack-reused 19
Receiving objects: 100% (102/102), 70.32 KiB | 0 bytes/s, done.
Resolving deltas: 100% (55/55), completed with 49 local objects.
From github.com:kubernetes/kompose
 * [new ref]         refs/pull/846/head -> pr_846
Switched to branch 'pr_846'

github.com/kubernetes/kompose  pr_846 ✔                                                                                                                                                                                                                                      2d  
▶ vim

github.com/kubernetes/kompose  pr_846 ✗                                                                                                                                                                                                                                    2d ⚑  
▶ make bin
go build -ldflags="-w -X github.com/kubernetes/kompose/cmd.GITCOMMIT=efb7552" -o kompose main.go

github.com/kubernetes/kompose  pr_846 ✗                                                                                                                                                                                                                                    2d ⚑  
▶ 
```

surajnarwade commented:
@cdrage after renaming it works locally but not in CI
after investing I found out that, we have to update `openshift` from vendor, which may cause trouble

cdrage commented:
Please troubleshoot and see if it works then.

On Nov 8, 2017 03:38, "Suraj Narwade" <notifications@github.com> wrote:

> @cdrage <https://github.com/cdrage> after renaming it works locally but
> not in CI
> after investing I found out that, we have to update openshift from
> vendor, which may cause trouble
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kompose/pull/846#issuecomment-342747897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGH-oCeoj1czJ95TUd_eNwstZ--t20DYks5s0WiQgaJpZM4P1BjX>
> .
>

cdrage commented on glide.yaml:
```diff
@@ -24,15 +24,15 @@ import:
 # We use libcompose to parse v1 and v2 of Docker Compose
 - package: github.com/docker/libcompose
-  version: 4a647d664afbe05c41455c9d534d8239671eb46a
+  version: 57bd716502dcbe1799f026148016022b0f3b989c
```
don't update libcompose, you don't need to.

surajnarwade replied:
Sirupsen :(

cdrage commented:
To give an update of what's happening after talking to @surajnarwade 

To update docker/cli we had to update vendoring. To update vendoring with the logrus errors, we had to update libcompose. To update libcompose, we had to update OpenShift vendoring due to conflicts, now we have to update and somehow migrate from 1.4 of OpenShift to 3.2 (I believe) in order to get all of this to work. :+1:

surajnarwade commented:
@cdrage @kadel instead of updating openshift, I have replaced `Sirupsen` with `sirupsen` in `docker/distribution`

cdrage commented:
@surajnarwade Can you please provide details on what you did?

Was it a PR that updated docker/distribution?

Did you manually edit the vendoring files?

Have you updated Makefile to sed replace Sirupsen to sirupsen, or anything else for that matter?

surajnarwade commented:
@cdrage @kadel m To give brief idea about this PR,

I have Updated `docker/cli` to get latest changes that @cdrage pushed for build v3 support

these changes comes with `sirupsen` case collsion problem, so updated `docker/libcompose` as well

after fixing `sirupsen` case collision, it again occured with package `docker/distribution` which is dependency of `openshift/origin`, since updating `openshift/origin` is too much tedious task.


To tackle this situation, I updated `imports` from `kompose` manually and  `docker/distribution` (from vendor) by adding line in Makefile and Now kompose seems to be working fine :)

I also added this thing to Makefile with `make vendor-update`

cdrage commented:
Awesome :+1: thanks for the detailed explanation!

One last thing, could you put the `make vendor-update` updates to Makefile into a separate commit so we can see what changes you've made? So in total, there should be three commits. One to the code, one to the Makefile and the other actually updating the vendoring files.

surajnarwade commented:
@cdrage gotcha

surajnarwade commented:
@cdrage now you can easily review it :)

cdrage commented:
@surajnarwade Your makefile commit is messy: https://github.com/kubernetes/kompose/pull/846/commits/c4044b56a56a1b6c34db88b35686555e0b8b5195

Seems like you included some vendor (libcompose) files and there are some code that you're modifying in `pkg`...

So to go into detail:
Commit 1: Updates to Kedge code (pkg, renaming sirupsen, etc.)
Commit 2: Updates to Makefile (and that's it!)
Commit 3: Any updates to vendoring file.

surajnarwade commented:
@cdrage , Updated commits, review needed

cdrage commented:
@surajnarwade tests aren't passing  on travis :(

surajnarwade commented:
@cdrage all green :)

cdrage commented on script/test/cmd/tests.sh:
```diff
@@ -84,6 +84,11 @@ cmd="kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-no
+# Test BuildConfig v3
+cmd="kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/docker-compose-v3.yml convert --stdout -j --build build-config"
+sed -e "s;%VERSION%;$version;g" -e "s;%CMD%;$cmd;g" -e "s;%URI%;$uri;g" -e "s;%REF%;$branch;g" $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/output-os-template-v3.json > /tmp/output-os.json
+convert::expect_success_and_warning "kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/docker-compose-v3.yml convert --stdout -j --build build-config" "/tmp/output-os.json" "$warning"
```
There should be another test added for Kubernetes.

cdrage commented:
Code / implementation LGTM, works on my side perfectly.

After merging, can you please update the documentation (conversion.md document) as well as open an issue that we need to add a test for testing Kubernetes build.

surajnarwade commented:
> Above works when I changed it to my username.

Thanks @cdrage  :)

Added test for kubernetes as well
Ready to merge :1st_place_medal:

cdrage commented:
Phew. Thanks A LOT for your contribution @surajnarwade  that was a lot of trouble-shooting. Merging in!

surajnarwade commented:
Thanks @cdrage
